### PR TITLE
Add Twirp support

### PR DESCRIPTION
**Pull request by austinylin**
This PR attempts to address #197 by adding support for [Twirp](https://github.com/twitchtv/twirp). This allows support for http/1.1 load balancers while leveraging protobuf and the automatic code generation that comes along with it.

Related PRs:
- go-kit support for Twirp: go-kit/kit#656
- fleet server support: kolide/fleet#1692

Still have some cleanup to do but wanted to get initial feedback.

## Todo
- [ ] Get go-kit support merged
- [ ] Cleanup Gopkg to remove overrides
- [ ] Add tests

### CODE REVIEW

File: Gopkg.lock [hunk 1]
```diff
@@ -55,10 +58,17 @@
...
-  version = "v0.6.0"
+  packages = [
+    "endpoint",
+    "log",
+    "log/level",
+    "transport/grpc",
+    "transport/twirp"
+  ]
+  revision = "bf0efd940670e10927665090e183b1aa65103cea"
+  source = "github.com/austinylin/kit"
```
Comment (marpaia) [inline Gopkg.lock]: So cool of you to add this to go-kit!

File: cmd/launcher/options.go [hunk 2]
```diff
@@ -113,6 +114,12 @@ func parseOptions() (*options, error) {
 			"Dial GRPC without a TLS config (default: false)",
 		)
 
+		flTransportTwirp = flag.Bool(
+			"transport_twirp",
```
Comment (marpaia) [inline cmd/launcher/options.go]: I think that this should be `--twirp` or `--transport=twirp`

File: cmd/launcher/launcher.go [hunk 3]
```diff
@@ -111,13 +111,25 @@ func main() {
...
 	defer db.Close()
 
-	conn, err := dialGRPC(opts.kolideServerURL, opts.insecureTLS, opts.insecureGRPC, logger)
-	if err != nil {
-		logger.Fatal("err", errors.Wrap(err, "dialing grpc server"))
+	var client service.KolideService
+	{
+		// configure the client with gRPC or Twirp.
+		useGRPC := !opts.transportTwirp
+		if useGRPC {
```
Comment (marpaia) [inline cmd/launcher/launcher.go]: Consider forgoing `useGRPC` and just do the following after inverting the loop:

```
if opts.transportTwirp {
```

**Comment by austinylin** [comment]
@marpaia updated the transport stuff to be more generic. Could you help me with the circleci failures? I can't figure out why it's failing to pull dependencies.

**Comment by zwass** [comment]
@austinylin I'm getting the following on my local machine:
```
make deps
go get -u github.com/golang/dep/cmd/dep
go get -u github.com/jteeuwen/go-bindata/...
dep ensure -vendor-only
dep: WARNING: branch, version, revision, or source should be provided for "github.com/boltdb/bolt"
dep: WARNING: branch, version, revision, or source should be provided for "github.com/kolide/kit"
dep: WARNING: branch, version, revision, or source should be provided for "github.com/kolide/osquery-go"
dep: WARNING: branch, version, revision, or source should be provided for "github.com/kolide/updater"
dep: WARNING: branch, version, revision, or source should be provided for "github.com/mixer/clock"
dep: WARNING: branch, version, revision, or source should be provided for "github.com/pkg/errors"
dep: WARNING: branch, version, revision, or source should be provided for "github.com/stretchr/testify"
dep: WARNING: branch, version, revision, or source should be provided for "golang.org/x/net"
dep: WARNING: branch, version, revision, or source should be provided for "golang.org/x/time"
dep: WARNING: branch, version, revision, or source should be provided for "google.golang.org/grpc"
dep: WARNING: branch, version, revision, or source should be provided for "github.com/twitchtv/twirp"
grouped write of manifest, lock and vendor: error while writing out vendor tree: failed to write dep tree: failed to export github.com/go-kit/kit: fatal: failed to unpack tree object bf0efd940670e10927665090e183b1aa65103cea
: exit status 128
make: *** [.deps] Error 1
```

This does not happen on `master`, so I'm trying to track down what the issue might be with the updated lockfile.

**Comment by marpaia** [comment]
I think this is failiing because the lockfile is pinning a commit hash which doesn't exist in the branch. `bf0efd940670e10927665090e183b1aa65103cea` is pinned, but if you look here (https://github.com/austinylin/kit/commits/twirp), I think you want `cb6ffcdc93fdc87130c9f0b252da3a775a1acdd7`.

**Comment by groob** [comment]
@austinylin the error that I'm getting is 
```
# github.com/kolide/launcher/simulator
simulator/simulator.go:364:27: undefined: service.New
# github.com/kolide/launcher/simulator
/tmp/go-build161499380/github.com/kolide/launcher/simulator/_test/_obj_test/simulator.go:372:27: undefined: service.New
# github.com/kolide/launcher/cmd/launcher
cmd/launcher/dial_test.go:77:12: undefined: service.New
cmd/launcher/dial_test.go:115:12: undefined: service.New
```

**Comment by marpaia** [comment]
From @groob to @austinylin in Slack on April 27, 2018 on the topic of landing this PR:

> Hey! Wanted to get your feedback. I was thinking of a few options. 
>
> 1) the go-kit PR has some feedback from the contributors. Idk what your thoughts/availability is to continue pushing it through code review. I’m pretty active in that space and wouldn’t mind opening a separate PR and driving it to a conclusion. But it’s your PR and I’d only want to do this if you’re too busy, uninterested or don’t care about getting the credit for the contribution :)
> 2) go-kit is just a bunch of libraries. To land support for twirp as a transport the go package itself doesn’t have to live within that namespace. we can keep your current but working prototype of the transport somewhere in a third party space. Then we can merge the transport options to fleet/launcher and refactor the transport further when the upstream changes are merged/final. This might also be a good option because we’d have some real feedback about usage to create consensus around the APIs. 
>
> Your PRs in launcher and fleet are good quality code and a lot of folks in the community have expressed interest in having these changes landed. So I’m happy to explore either of the above, or a third option.

**Comment by marpaia** [comment]
I'm closing this due to inactivity. @austinylin if you're able to circle back on this, we would love to re-open this PR and get this merged one of these days.

**Comment by hamstah** [comment]
Hi,
Are there any more recent plans to support http?
Thanks

**Comment by zwass** [comment]
Yes, Launcher supports http through `--transport=jsonrpc`.

**Comment by blaedj** [comment]
Note that while launcher supports http via the `--transport=jsonrpc` flag, [fleet](https://github.com/kolide/fleet) does **not** support the jsonrpc transport at this time.

**Pull request closed without merging**